contracts/src/v0.8/keystone/KeystoneForwarder.sol

```diff
@@ -11,66 +11,150 @@ contract KeystoneForwarder is IForwarder, ConfirmedOwner, TypeAndVersionInterfac
+      }
+  }
+
+  // send a report to receiver
   function report(
-    address targetAddress,
+    address receiver,
     bytes calldata data,
     bytes[] calldata signatures
   ) external nonReentrant returns (bool) {
-    if (data.length < SELECTOR_LENGTH + REPORT_LENGTH) {
+    if (data.length < SELECTOR_LENGTH + Utils.REPORT_HEADER_LENGTH) {
       revert InvalidData(data);
     }
 
     // data is an encoded call with the selector prefixed: (bytes4 selector, bytes report, ...)
     // we are able to partially decode just the first param, since we don't know the rest
+    bytes4 selector = getSelector(data);
     bytes memory rawReport = abi.decode(data[4:], (bytes));
 
-    // TODO: we probably need some type of f value config?
+    (/* bytes32 workflowId */, bytes4 donId, bytes32 workflowExecutionId) = Utils._splitReport(rawReport);
+
+    if (signatures.length != s_configs[donId].f + 1) {
+      revert WrongNumberOfSignatures();
+    }
+
+    // f can never be 0, so this means the config doesn't actually exist
+    if (s_configs[donId].f == 0) revert InvalidDonId();
 
     bytes32 hash = keccak256(rawReport);
 
     // validate signatures
+    address[MAX_ORACLES] memory signed;
+    uint8 index = 0;
     for (uint256 i = 0; i < signatures.length; i++) {
-      // TODO: is libocr-style multiple bytes32 arrays more optimal?
+      // TODO: is libocr-style multiple bytes32 arrays more optimal, gas-wise?
       (bytes32 r, bytes32 s, uint8 v) = Utils._splitSignature(signatures[i]);
       address signer = ecrecover(hash, v, r, s);
-      // TODO: we need to store oracle cluster similar to aggregator then, to validate valid signer list
+
+      // validate signer is trusted and signature is unique
+      index = uint8(s_configs[donId]._positions[signer]);
+      if (index == 0) revert InvalidSigner(); // index is 1-indexed so we can detect unset signers
+      index -= 1;
+      if(signed[index] != address(0)) revert RepeatedSigner();
+      signed[index] = signer;
     }
 
-    (bytes32 workflowId, bytes32 workflowExecutionId) = Utils._splitReport(rawReport);
+    bytes32 reportId = _reportId(receiver, selector, workflowExecutionId);
 
-    // report was already processed
-    if (s_reports[workflowExecutionId] != address(0)) {
+    if (s_reports[reportId] != address(0)) {
+      // report was already processed
       return false;
     }
 
     // solhint-disable-next-line avoid-low-level-calls
-    (bool success, bytes memory result) = targetAddress.call(data);
+    (bool success, bytes memory result) = receiver.call(data);
 
-    s_reports[workflowExecutionId] = msg.sender;
+    s_reports[reportId] = msg.sender;
     return true;
   }
 
+  function _reportId(address receiver, bytes4 selector, bytes32 workflowExecutionId) internal pure returns (bytes32) {
+    // TODO: gas savings: could we just use a bytes key and avoid another keccak256 call
+    return keccak256(bytes.concat(bytes20(uint160(receiver)), selector, workflowExecutionId));
```

Comment:
```suggestion
    return keccak256(abi.encodePacked(receiver, selector, workflowExecutionId);
```
Also, see comment about selector - I think it should be fixed, so it doesn't need to be in the reportId